man/go-camo.1.scd
go-camo(1)

# NAME

go-camo - Go version of Camo server

# SYNOPSIS

*go-camo* [<_OPTION_> [<_OPTION-ARGUMENT_>]]...

# DESCRIPTION

_go-camo_(1) is an implementation of Camo in Go.

Camo is a special type of image proxy that proxies non-secure images over SSL/TLS.++
This prevents mixed content warnings on secure pages.

It works in conjunction with back-end code to rewrite image URLs and sign them
with an HMAC.

# ENVIRONMENT VARS

*GOCAMO_HMAC*
	The HMAC key to use.

*HTTPS_PROXY*
	Configure an outbound proxy that will be used as the proxy URL for HTTPS
	requests.

	Either a complete URL, or a host[:port] (in which case the _http_ scheme is assumed).

*HTTP_PROXY*
	Configure an outbound proxy that will be used as the proxy URL for HTTP
	requests.

	Either a complete URL, or a host[:port] (in which case the _http_ scheme is assumed).

Additional environment vars are show as part of the `--help` output.

Regarding *Environmen Vars* vs *OPTIONS* precedence: _OPTIONS_, if provided,
override those defined in environment variables.++
For exmaple, if the HMAC key is provided on the command line, it will override
(if present), an HMAC key set in the environment var.

# OPTIONS

*-k*, *--key*=<_HMAC_KEY_>
	The HMAC key to use.

*-H*, *--header*=<_HEADER_>
	Add additional header to each response.

	This option can be used multiple times to add multiple headers.++
	When specified, these headers are set unconditionally on all responses.

	See _ADD_HEADERS_ for more info.

*--listen*=<_ADDRESS:PORT_>
	Address and port to listen to, as a string of _ADDRESS:PORT_.++
	Default: 0.0.0.0:8080

*--socket-listen*=<_SOCKET_PATH_>
	Path for unix domain socket to bind to for HTTP.

*--ssl-listen*=<_ADDRESS:PORT_>
	Address and port to listen via SSL to, as a string of _ADDRESS:PORT_.

*--ssl-key*=<_SSL-KEY-FILE_>
	Path to ssl private key.++
	Default: key.pem

*--ssl-cert*=<_SSL-CERT-FILE_>
	Path to ssl certificate.++
	Default: cert.pem

*--quic*
	Enable http3/quic. Binds to the same port number as _--ssl-listen_ but udp+quic.

*--automaxprocs*
	Set GOMAXPROCS automatically to match Linux container CPU quota/limits.

*--max-size*=<_SIZE_>
	Max response size allowed in KB. Set to 0 to disable size restriction.++
	Default: 0

*--max-size-redirect*=<_URL_>
	URL to redirect to when max-size is exceeded.

*--timeout*=<_TIME_>
	Timeout value for upstream response. Format is "4s" where s means seconds (backend).++
	Default: 4s

*--idletimeout*=<_TIME_>
	Maximum amount of time to wait for the next request when keep-alive is enabled (frontend).++
	Default: 60s

*--readtimeout*=<_TIME_>
	Maximum duration for reading the entire request, including the body (frontend).++
	Default: 30s

*--max-redirects*
	Maximum number of redirects to follow.++
	Default: 3

*--metrics*
	Enable Prometheus compatible metrics endpoint.

	If the _metrics_ flag is provided, then the service will expose a Prometheus
	/metrics endpoint and a /debug/vars endpoint from the go _expvar_ package.

	See _METRICS_ for more info.

*--no-debug-vars*
	If the _no-debug-vars_ flag is provided along with the _metrics_ flag, the
	/debug/vars endpoint is removed.

	See _METRICS_ for more info.

*--no-log-ts*
	Do not add a timestamp to logging output.

*--no-fk*
	Disable frontend http keep-alive support (frontend).

*--no-bk*
	Disable backend http keep-alive support (backend).

*--log-json*
	Log in JSON format.

*--allow-content-video*
	Additionally allow video/\* content type.

*--allow-content-audio*
	Additionally allow audio/\* content type.

*--allow-credential-urls*
	Allow urls to contain user/pass credentials.

*--filter-ruleset*=<_FILE_>
	Path to a text file that contains a list (one per line) filter rules.

	If an filter-ruleset file is defined, that file is read and each line
	converted into a filter rule.++
	If the request fails the rule-set, the request is denied.

	See _go-camo-filtering_(5) for more information.

*--server-name*=<_SERVER-NAME_>
	Value to use for the HTTP server field.++
	Default: go-camo

*--user-agent*=<_USER-AGENT_>
	Value to use for the User-Agent http request header.++
	Default: go-camo

*--expose-server-version*
	Include the server version in the HTTP server response header.

*--enable-xfwd4*
	Enable x-forwarded-for passthrough/generation.

*-v*, *--verbose*
	Show verbose (debug) level log output

*-V*, *--version*
	Print version and exit; specify twice to show license information.

*-h*, *--help*
	Show help output and exit.


# ADD_HEADERS

Additional default headers (headers sent on every reply) can be set with the
*-H, --header* flag. This option can be used multiple times.

The list of default headers sent are:

```
X-Content-Type-Options: nosniff
X-XSS-Protection: 1; mode=block
Content-Security-Policy: default-src 'none'
```

Additional headers are added to the above set.

As an example, if you wanted to return an Strict-Transport-Security and an
X-Frame-Options header by default, you could add this to the command line:

```
go-camo -k BEEFBEEFBEEF \\
    -H "Strict-Transport-Security: max-age=16070400" \\
    -H "X-Frame-Options: deny"
```

# METRICS

When the *--metrics* flag is used, the service will expose a
Prometheus-compatible /metrics endpoint. This can be used by monitoring
systems to gather data.

The endpoint includes all of the default go_ and process_. In addition, a
number of custom metrics.

A /debug/vars endpoint is also included with *--metrics* by default.
This endpoint returns memstats and some additional data. This endpoint can be
disabled by additionally supplying the *--no-debug-vars* flag.

[- *Metric*
:- *Type*
:- *Description*
|[ camo_response_duration_seconds
:- Histogram
:< A histogram of latencies for proxy responses.
|  camo_response_size_bytes
:  Histogram 
:  A histogram of sizes for proxy responses.
|  camo_proxy_content_length_exceeded_total
:  Counter
:  The number of requests where the content length was exceeded.
|  camo_proxy_reponses_failed_total
:  Counter
:  The number of responses that failed to send to the client.
|  camo_proxy_reponses_truncated_total
:  Counter
:  The number of responess that were too large to send.
|  camo_responses_total
:  Counter
:  Total HTTP requests processed by the go-camo, excluding scrapes.


It also includes a _camo_build_info_ metric that exposes the version. In
addition, you can expose some extra data to metrics via env vars, if desired:

- Revision via _APP_INFO_REVISION_
- Branch via _APP_INFO_BRANCH_
- BuildDate via _APP_INFO_BUILD_DATE_
- You can also override the version by setting _APP_INFO_VERSION_

# EXAMPLES

Listen on loopback port 8080 with a upstream timeout of 6 seconds:

```
go-camo -k BEEFBEEFBEEF \\
    --listen=127.0.0.1:8080 \\
    --timeout=6s
```

Set HMAC key via env var, and an HSTS header:

```
export GOCAMO_HMAC=BEEFBEEFBEEF
go-camo \\
    --listen=127.0.0.1:8080 \\
    --timeout=6s \\
    -H "Strict-Transport-Security: max-age=16070400"
```

# WEBSITE

https://github.com/cactus/go-camo

# SEE_ALSO

https://github.com/atmos/camo


CHANGELOG.md
# HEAD

# v3.0.0
- When configuring a max-size, this will now also disable backend http keep-alives, to
  ensure that the connection can be closed before fully reading the response (prevents
  resource exhaustion).  
  ref: [go-1.27 release notes](https://go.dev/doc/go1.27#:~:text=HTTP/1%20Response.Body,likely%20be%20beneficial.)
- add more comprehensive support for cli flags to be set via env vars, with names prefixed with `GOCAMO_`.
  Use `--help` to see names.
- upgrade go version to 1.27
- vendor go dependencies
- drop treeprint from dependency (reimplemented basic version locally)
- include codeberg.org/dropwhile/assert in pkg dir directly, remove as mod dep
- handle github.com/prometheus/common/version registration directly, remove as
  top level mod dep

# v2.7.5 2026-07-08
- bump dependencies
- build binaries with go-1.26.5

# v2.7.4 2026-04-25
- bump dependencies
- build binaries with go-1.26.2

# v2.7.3 2026-02-15

- bump dependencies
- build binaries with go-1.26.0  
  refs:  
  * https://tip.golang.org/doc/go1.26
  * https://antonz.org/go-1-26/

# v2.7.2 2025-12-08

- bump dependencies
- build binaries with go-1.25.5

# v2.7.1 2025-10-08

- bump dependencies
- build binaries with go-1.25.2

# v2.7.0 2025-10-01

- fix: add image/svg+xml to acceptTypes. (#84)
- add ability to set user-agent, distinct from server-name

# v2.6.5 2025-08-16

- rebuild binaries with go-1.25
- remove gomaxprocs dependencies, as this functionality now natively handled by
  Go itself (as of go-1.25)

# v2.6.4 2025-08-09

- Support for ECS quota detection with automaxprocs  
  ref: [\#83](https://github.com/cactus/go-camo/pull/83)

- Update depencencies

# v2.6.3 2025-04-01

- Build with go-1.24.2

- Update depencencies

# v2.6.2 2025-02-22

- Switch man-page generate to scdoc.

- Update dependencies

- Build with go-1.24

# v2.6.1 2024-12-19

- Update dependencies

# v2.6.0 2024-08-31

- Start building with Go-1.23.0

- Add CLI flag to expose ReadTimeout for tuning (frontend).  
  Default is 30s (same as previosly).

- Add CLI flag for configuring IdleTimeout, the maximum amount of time
  to wait for the next request when keep-alive is enabled (frontend).  
  Default is 30 seconds.

  > [!Note]
  > Previously this value was not set, so it defaulted to the zero value,
  > which resulted in the ReadTimeout value (default 30s) being used
  > internally by the go http server. The new default mirrors this
  > previous old default.

# v2.5.1 2024-07-02

- Rebuild release with Go-1.22.5

# v2.5.0 2024-06-27

- Implement optional redirect when max-size is exceeded  
  ref: [\#80](https://github.com/cactus/go-camo/pull/80)

# v2.4.13 2024-04-22

- Release tagged for arm64 docker building only.

# v2.4.12 2024-04-20

- Update library dependencies.

- Fix docker and github packages publishing issue.

# v2.4.11 2024-04-03

- Update library dependencies.

- Build with Go-1.22.2

# v2.4.10 2024-03-17

- Update library dependencies.

# v2.4.9 2024-02-16

- Minimum Go version now 1.21 due to quic-go dependency, due to better
  cryto/tls support for QUIC in Go-1.21.

- Update library dependencies.

# v2.4.8 2023-12-19

- Add `--automaxprocs` flag to set GOMAXPROCS automatically to match
  Linux container CPU quota/limits.

- Update library dependencies.

# v2.4.7 - 2023-11-13

- Add http3/quic server support. New flag `--quic`. Requires
  `--ssl-listen`.

# v2.4.6 - 2023-10-25

- Add `--no-debug-vars` flag to disable /debug/vars when `--metrics` is
  enabled. (#66, \#67)

# v2.4.5 - 2023-10-23

- fix htrie matching of non punycode (eg. unicode) idna hostnames

- slightly faster logging (update to mlog dependency)

- address a logging issue with missing url path output in
  `"built outgoing request"` debug log

- moderate improve performance of hostname rule processing (approx
  12-30% in microbenchmarks)

- slight improvement in request path url processing (approx 2-4% in
  microbenchmarks)

- fix /debug/vars being enabled by default (#65) due to expvars import
  side effect

# v2.4.4 - 2023-07-25

- update dependencies

- bump version in go.mod (and fix all internal module references)  
  ref: discussion
  [\#62](https://github.com/cactus/go-camo/discussions/62)

# v2.4.3 - 2023-02-18

- update library dependency golang.org/x/net.  
  refs:
  [dependabot-3](https://github.com/cactus/go-camo/security/dependabot/3),
  [dependabot-4](https://github.com/cactus/go-camo/security/dependabot/4)

# v2.4.2 - 2023-02-16

- update library dependency prometheus, covering CVE-2022-21698.  
  Note that for go-camo, the issue in the prometheus library was
  exploitable only when the metrics option/flag (--metrics) is enabled.

- build with go1.19.5

# v2.4.1 - 2022-09-28

- Rebuild release with go-1.19.1

# v2.4.0 - 2022-01-30

- Add support for internal address proxies (HTTP(S)\_PROXY).  
  issue \#55

- Add some additional documentation/caveats regarding HTTP(S)\_PROXY
  usage.

# v2.3.0 - 2021-10-20

- Add support for listening on a unix socket.

- A more graceful shutdown process, under SIGINT or SIGTERM.

# v2.2.2 - 2021-08-21

- Change test only helper library

- structure logging (json) support for http requests (debug logging).  
  PR \#52

# v2.2.1 - 2021-03-25

- Update some dependencies.

# v2.2.0 - 2021-01-10

- Move ip filtering to Dialer.Control, to further improve SSRF
  protections.  
  Note: Added a few additional debug log messages emitted when ip
  filtering

# v2.1.5 - 2020-08-11

- Rebuild release with go-1.15

# v2.1.4 - 2020-02-18

- Rebuild release with go-1.13.8

- Experimental windows build

# v2.1.3 - 2019-12-18

- Rebuild release with go-1.13.5

# v2.1.2 - 2019-11-17

- Fix for enabling metrics collection in proxy

# v2.1.1 - 2019-11-12

- Security fixes / content-type validation

- Add `ProxyFromEnvironment` support. This uses HTTP proxies directed by
  the `HTTP_PROXY` and `NO_PROXY` (or `http_proxy` and `no_proxy`)
  environment variables. See

src/go-camo/go-camo.go
// Copyright (c) 2012-2016 Eli Janssen
// Use of this source code is governed by an MIT-style
// license that can be found in the LICENSE file.

// go-camo daemon (go-camod)
package main

import (
	"fmt"
	"io/ioutil"
	"log"
	"net/http"
	"os"
	"runtime"
	"strconv"
	"strings"
	"syscall"
	"time"

	"go-camo/camo"
	"go-camo/router"
	"go-camo/stats"

	"github.com/cactus/gologit"
	flags "github.com/jessevdk/go-flags"
)

var (
	ServerName    = "go-camo"
	ServerVersion = "no-version"
)

func main() {
	var gmx int
	if gmxEnv := os.Getenv("GOMAXPROCS"); gmxEnv != "" {
		gmx, _ = strconv.Atoi(gmxEnv)
	} else {
		gmx = runtime.NumCPU()
	}
	runtime.GOMAXPROCS(gmx)

	// command line flags
	var opts struct {
		HMACKey             string        `short:"k" long:"key" description:"HMAC key"`
		AddHeaders          []string      `short:"H" long:"header" description:"Extra header to return for each response. This option can be used multiple times to add multiple headers"`
		Stats               bool          `long:"stats" description:"Enable Stats"`
		AllowList           string        `long:"allow-list" description:"Text file of hostname allow regexes (one per line)"`
		MaxSize             int64         `long:"max-size" default:"5120" description:"Max response image size (KB)"`
		ReqTimeout          time.Duration `long:"timeout" default:"4s" description:"Upstream request timeout"`
		MaxRedirects        int           `long:"max-redirects" default:"3" description:"Maximum number of redirects to follow"`
		DisableKeepAlivesFE bool          `long:"no-fk" description:"Disable frontend http keep-alive support"`
		DisableKeepAlivesBE bool          `long:"no-bk" description:"Disable backend http keep-alive support"`
		BindAddress         string        `long:"listen" default:"0.0.0.0:8080" description:"Address:Port to bind to for HTTP"`
		BindAddressSSL      string        `long:"ssl-listen" description:"Address:Port to bind to for HTTPS/SSL/TLS"`
		SSLKey              string        `long:"ssl-key" description:"ssl private key (key.pem) path"`
		SSLCert             string        `long:"ssl-cert" description:"ssl cert (cert.pem) path"`
		Verbose             bool          `short:"v" long:"verbose" description:"Show verbose (debug) log level output"`
		Version             bool          `short:"V" long:"version" description:"print version and exit"`
	}

	// parse said flags
	_, err := flags.Parse(&opts)
	if err != nil {
		if e, ok := err.(*flags.Error); ok {
			if e.Type == flags.ErrHelp {
				os.Exit(0)
			}
		}
		os.Exit(1)
	}

	if opts.Version {
		fmt.Printf("%s %s (%s,%s-%s)\n", ServerName, ServerVersion, runtime.Version(), runtime.Compiler, runtime.GOARCH)
		os.Exit(0)
	}

	config := camo.Config{}
	if hmacKey := os.Getenv("GOCAMO_HMAC"); hmacKey != "" {
		config.HMACKey = []byte(hmacKey)
	}

	// flags override env var
	if opts.HMACKey != "" {
		config.HMACKey = []byte(opts.HMACKey)
	}

	if len(config.HMACKey) == 0 {
		log.Fatal("HMAC key required")
	}

	if opts.BindAddress == "" && opts.BindAddressSSL == "" {
		log.Fatal("One of bind-address or bind-ssl-address required")
	}

	if opts.BindAddressSSL != "" && opts.SSLKey == "" {
		log.Fatal("ssl-key is required when specifying bind-ssl-address")
	}
	if opts.BindAddressSSL != "" && opts.SSLCert == "" {
		log.Fatal("ssl-cert is required when specifying bind-ssl-address")
	}

	// set keepalive options
	config.DisableKeepAlivesBE = opts.DisableKeepAlivesBE
	config.DisableKeepAlivesFE = opts.DisableKeepAlivesFE

	if opts.AllowList != "" {
		b, err := ioutil.ReadFile(opts.AllowList)
		if err != nil {
			log.Fatal("Could not read allow-list", err)
		}
		config.AllowList = strings.Split(string(b), "\n")
	}

	AddHeaders := map[string]string{
		"X-Content-Type-Options":  "nosniff",
		"X-XSS-Protection":        "1; mode=block",
		"Content-Security-Policy": "default-src 'none'",
	}

	for _, v := range opts.AddHeaders {
		s := strings.SplitN(v, ":", 2)
		if len(s) != 2 {
			log.Printf("ignoring bad header: '%s'\n", v)
			continue
		}

		s0 := strings.TrimSpace(s[0])
		s1 := strings.TrimSpace(s[1])

		if len(s0) == 0 || len(s1) == 0 {
			log.Printf("ignoring bad header: '%s'\n", v)
			continue
		}
		AddHeaders[s[0]] = s[1]
	}

	// convert from KB to Bytes
	config.MaxSize = opts.MaxSize * 1024
	config.RequestTimeout = opts.ReqTimeout
	config.MaxRedirects = opts.MaxRedirects
	config.ServerName = ServerName

	// set logger debug level and start toggle on signal handler
	logger := gologit.Logger
	logger.Set(opts.Verbose)
	logger.Debugln("Debug logging enabled")
	logger.ToggleOnSignal(syscall.SIGUSR1)

	proxy, err := camo.New(config)
	if err != nil {
		log.Fatal(err)
	}

	dumbrouter := &router.DumbRouter{
		ServerName:  config.ServerName,
		AddHeaders:  AddHeaders,
		CamoHandler: proxy,
	}

	if opts.Stats {
		ps := &stats.ProxyStats{}
		proxy.SetMetricsCollector(ps)
		log.Println("Enabling stats at /status")
		dumbrouter.StatsHandler = stats.StatsHandler(ps)
	}

	http.Handle("/", dumbrouter)

	if opts.BindAddress != "" {
		log.Println("Starting server on", opts.BindAddress)
		go func() {
			srv := &http.Server{
				Addr:        opts.BindAddress,
				ReadTimeout: 30 * time.Second}
			log.Fatal(srv.ListenAndServe())
		}()
	}
	if opts.BindAddressSSL != "" {
		log.Println("Starting TLS server on", opts.BindAddressSSL)
		go func() {
			srv := &http.Server{
				Addr:        opts.BindAddressSSL,
				ReadTimeout: 30 * time.Second}
			log.Fatal(srv.ListenAndServeTLS(opts.SSLCert, opts.SSLKey))
		}()
	}

	// just block. listen and serve will exit the program if they fail/return
	// so we just need to block to prevent main from exiting.
	select {}
}
